core/vnl/vnl_na.cxx

```diff
@@ -84,32 +85,68 @@ float vnl_na_nan_to_na(float v)
   return vnl_math::isnan(v) ? vnl_na(float()) : v;
 }
 
-
 //: Read a floating point number or "NA" from a stream.
-template <class T> inline void vnl_na_extract_type(vcl_istream &is, T& x)
+template <class T> inline void vnl_na_extract_type(vcl_istream &is, T& value)
 {
   if (!is) return;
-  is >> x;
-
-  if (!!is || is.eof()) return;
-  is.clear();
-
-  char c=' ';
-  is >> c;
-  if (c != 'N' && c!='n')
-  {
-    is.putback(c);
-    is.clear(vcl_ios::badbit);
-    return;
-  }
-  is >> c;
-  if (c != 'A' && c!='a')
-  {
-    is.putback(c);
-    is.clear(vcl_ios::badbit);
-    return;
+  vcl_stringstream oneToken("");
+  unsigned int char_processed_count = 0;
+  bool period_found = false;
+  bool current_location_is_delimiter = false;
+
+  while (!is.eof()) {
+    vcl_stringstream::char_type c;
+    vcl_istream::int_type p = is.peek();
+    if ( char_processed_count == 0 ) { //The first character is the start of the token of interest.
+      if (std::isspace(p)) {
+        is.get(c); // Gobble up the peeked at character
+        continue; //Gobble up preceeding white space
+      }
+      if ( p == 'N' || p == 'n' ) {
+        is.get(c);// Gobble up the N
+        p = is.peek();
+        if (p == 'A' || p == 'a') {
+          is.get(c); // Gobble up the A
+          value = vnl_na(T());
+          return;
+        }
+        else
+        {
+          std::string checkForNAString;
+          is >> checkForNAString; //Gobble up the rest of the token, whatever that is
+          value = 999.999; // Invalid parsing occured
+          return;
+        }
+      }
+    }
+    // Find if character is candidate for float values
+    if (std::isdigit(p) || p == '-' || p == '+' || p == '.') {
```

Comment:
Use `vcl_isdigit` instead of `std::isdigit`.  Include `vcl_cctype.h` at the top to get it.